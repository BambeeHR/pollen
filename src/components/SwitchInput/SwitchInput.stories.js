import BaseIcon from '../BaseIcon/BaseIcon.vue';
import Form from '../../constants/Form';
import SwitchInput from './SwitchInput.vue';
import TypeOverline from '../TypeOverline/TypeOverline.vue';

const sizes = [Form.Sizes.DENSE, Form.Sizes.NORMAL];

export default {
  title: 'Toggles/SwitchInput',
  component: SwitchInput,
  parameters: {
    componentSubtitle: 'A single checkbox input styled as a switch',
  },
};

export const Gallery = () => ({
  components: { SwitchInput, TypeOverline },
  data() {
    return {
      sizes,
      value: true,
    };
  },
  template: `
  <div>
    <div v-for="size in sizes" :key="size">
      <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My checkbox" />
      <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My checkbox" invalid />
      <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My checkbox" disabled />
    </div>
  </div>
  `,
});

export const LabelOnTheLeft = () => ({
  components: { SwitchInput, TypeOverline },
  data() {
    return {
      sizes,
      value: true,
    };
  },
  template: `
  <div>
    <div v-for="size in sizes" :key="size">
      <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My checkbox" label-position="left" />
      <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My checkbox" label-position="left" invalid />
      <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My checkbox" label-position="left" disabled />
    </div>
  </div>
  `,
});

export const HTMLLabel = () => ({
  components: { SwitchInput, TypeOverline },
  data() {
    return {
      sizes,
      value: true,
    };
  },
  template: `
  <div>
    <div v-for="size in sizes" :key="size">
      <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My <i>checkbox</i>" />
      <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My <i>checkbox</i>" invalid />
      <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
      <SwitchInput v-model="value" :size="size" label="My <i>checkbox</i>" disabled />
    </div>
  </div>
  `,
});

export const WithLabelAsSlot = () => ({
  components: { BaseIcon, SwitchInput, TypeOverline },
  data() {
    return {
      sizes,
      value: true,
    };
  },
  template: `
  <div>
    <div v-for="size in sizes" :key="size">
      <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
      <SwitchInput v-model="value" :size="size">
        <div>
          <BaseIcon icon="lock" />
          Lock it down?
        </div>
      </SwitchInput>
      <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
      <SwitchInput v-model="value" :size="size" invalid>
        <div>
          <BaseIcon icon="lock" />
          Lock it down?
        </div>
      </SwitchInput>
      <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
      <SwitchInput v-model="value" :size="size" disabled>
        <div>
          <BaseIcon icon="lock" />
          Lock it down?
        </div>
      </SwitchInput>
    </div>
  </div>
  `,
});

/* eslint-disable no-param-reassign */
[Gallery, LabelOnTheLeft, HTMLLabel, WithLabelAsSlot].forEach((item) => {
  item.story = {
    parameters: {
      jest: ['SwitchInput.test.js'],
    },
  };
});
