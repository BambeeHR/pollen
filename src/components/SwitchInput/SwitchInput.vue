<template>
  <ToggleInput
    :id="id"
    class="switch-input"
    type="checkbox"
    :checked="value === trueValue"
    :disabled="disabled"
    :invalid="invalid"
    :label="label"
    :label-position="labelPosition"
    :name="name || id"
    :size="size"
    variant="switch"
    @click="$emit('input', $event.target.checked ? trueValue : falseValue)"
  >
    <slot />
  </ToggleInput>
</template>

<script>
import Form from '../../constants/Form';
import ToggleInput from '../internal/forms/toggles/ToggleInput.vue';

/**
 * `import { SwitchInput } from '@bambeehr/pollen';`
 *
 * A single checkbox input that presents as an on/off switch. This is compatible
 * with v-model, including using the `true-value` and `false-value` props.
 * https://vuejs.org/v2/guide/forms.html#Checkbox-1
 */
export default {
  components: { ToggleInput },
  props: {
    /** If the input is disabled. */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Either a string or boolean. The value to be emitted on an input event
     * when the input is unchecked. Compatible with `v-model` and how normal
     * checkbox inputs work within Vue.
     */
    falseValue: {
      type: [String, Boolean],
      default: false,
    },
    /** A unique element ID. By default, one is randomly generated. */
    id: {
      type: String,
      default: null,
    },
    /** If true, this field will display in an error state. */
    invalid: {
      type: Boolean,
      default: false,
    },
    /**
     * A label to display next to the field. Alternately, the default `slot`
     * can be used to display more complex labels.
     */
    label: {
      type: String,
      default: null,
    },
    /**
     * Where the label should be positioned relative to the toggle, one of
     * `left` or `right`.
     */
    labelPosition: {
      type: String,
      default: Form.LabelPositions.RIGHT,
      validator: (value) => Object.values(Form.LabelPositions).includes(value),
    },
    /**
     * A name attribute. Falls back to `id` if none is supplied. If neither is
     * supplied, one is randomly generated.
     */
    name: {
      type: String,
      default: null,
    },
    /** One of `dense`, or `regular`. */
    size: {
      type: String,
      default: Form.Sizes.NORMAL,
      validator: (value) =>
        [Form.Sizes.DENSE, Form.Sizes.NORMAL].includes(value),
    },
    /**
     * Either a string or boolean. The value to be emitted on an input event
     * when the input is checked. Compatible with `v-model` and how normal
     * checkbox inputs work within Vue.
     */
    trueValue: {
      type: [String, Boolean],
      default: true,
    },
    /**
     * Either a string or boolean. Compatible with v-model. Checkboxes are
     * considered checked when `value === trueValue`.
     */
    value: {
      type: [String, Boolean],
      default: false,
    },
  },
};
</script>
