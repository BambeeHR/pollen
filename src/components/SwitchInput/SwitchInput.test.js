import { mount } from '@vue/test-utils';
import SwitchInput from './SwitchInput.vue';

describe('SwitchInput', () => {
  test('renders correctly', () => {
    const wrapper = mount(SwitchInput, {
      propsData: {
        value: true,
        id: 'test',
        label: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('emits an input event', async () => {
    const Component = {
      components: { SwitchInput },
      template: `
        <div>
          <SwitchInput v-model="value" label="test" />
        </div>
      `,
      data() {
        return {
          value: false,
        };
      },
    };
    const wrapper = mount(Component);
    const input = wrapper.find('input');
    await input.trigger('click');
    expect(wrapper.vm.value).toEqual(true);
    await input.trigger('click');
    expect(wrapper.vm.value).toEqual(false);
  });

  test('emits an input event with defined trueValue and falseValue prop strings', async () => {
    const Component = {
      components: { SwitchInput },
      template: `
        <div>
          <SwitchInput v-model="value" true-value="yes" false-value="no" label="test" />
        </div>
      `,
      data() {
        return {
          value: 'yes',
        };
      },
    };
    const wrapper = mount(Component);
    const input = wrapper.find('input');
    await input.trigger('click');
    expect(wrapper.vm.value).toEqual('no');
    await input.trigger('click');
    expect(wrapper.vm.value).toEqual('yes');
  });
});
