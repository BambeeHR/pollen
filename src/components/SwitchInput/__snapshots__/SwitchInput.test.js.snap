// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchInput renders correctly 1`] = `
<label
  class="toggle-input switch-input toggle-input--normal toggle-input--switch toggle-input--label-right"
  for="test"
>
  <input
    class="toggle-input__input"
    id="test"
    name="test"
    type="checkbox"
  />
   
  <div
    class="switch-indicator switch-indicator--normal switch-indicator--checked"
    variant="switch"
  />
   
  <div
    class="toggle-input__label"
  >
    <span>
      test
    </span>
  </div>
</label>
`;
